TGUI say: Upgrades chat input with modern features

S34NW opened the pull request:
<!-- By ticking or leaving ticked the option "Allow edits and access to secrets by maintainers" you give permission for repository maintainers to push changes to your branch without explicitly asking. -->

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
Ports the following PRs:
https://github.com/tgstation/tgstation/pull/67116
https://github.com/tgstation/tgstation/pull/73212
https://github.com/tgstation/tgstation/pull/75431
https://github.com/tgstation/tgstation/pull/75534

- Creates a super streamlined tgui-say package that's independent of TGUI
- Adds forced speech mechanics



<details>
  <summary>Click me for feature previews</summary>

window autoexpands on entry
![auto expand](https://user-images.githubusercontent.com/42397676/173295383-d2978988-c8f5-4a50-a179-cc5eadcf7a09.gif)

five message history: use arrows up/down
![chat history](https://user-images.githubusercontent.com/42397676/173777246-a3321469-5ab2-445a-abb2-66d5b7490678.gif)

light mode
![lightmode](https://user-images.githubusercontent.com/42397676/173777299-54821553-6c27-434c-b3ed-0c12a3cad0dc.png)
 
change channels by TAB/button OR typing department into me/radio
![channels](https://user-images.githubusercontent.com/42397676/173295417-85c38cc2-cfe1-4de5-aadc-ffcf597f56bf.gif)

quick radio hotkey "Y"
![radio hotkey](https://user-images.githubusercontent.com/42397676/173295481-e67879e2-165f-4145-87df-dfcbccce4650.gif)

thinking (window open) and typing indicators with inactivity detection note: indicators can be turned off, do not show up for "secret" channels like OOC and binary
![typing indicators](https://user-images.githubusercontent.com/42397676/173295577-f3c6f2a6-1f90-4df7-b8fb-8db80f4a3586.gif)
</details>
<!-- Include a small to medium description of what your PR changes. -->
<!-- Document all changes, as not doing this may delay reviews or even discourage maintainers from merging your PR! -->
<!-- If your PR fixes an issue, add "Fixes #1234" somewhere in the PR description. This will automatically close the bug upon PR submission. -->

## Why It's Good For The Game
Finally we shirk off the most tenacious part of this 1996 UI with a sleek and modern replacement that's CUSTOMIZABLE
<!-- Add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Testing
Talking, hitting myself, etc.
<!-- How did you test the PR, if at all? -->

## Changelog
:cl: S34NW Aylong220 GDN jlsnow301 stylemistake
add: Adds the TGUI say modal to replace speech input boxes. Many features!
add: You can switch output channel within the TGUI say modal by clicking the channel button or keying TAB.
add: Added 5-message chat history to the say window. Use arrow keys to scroll recent messages.
add: The say window auto-expands! It adds up to two extra lines and scrolls beyond that.
add: The common radio hotkey - "Y". No need to type ; in radio channel. Help maint!
add: Radio subchannel helpers - Typing subchannels like ":e " will show labels.
add: Themes for TGUI Say! There is currently a light mode. More to come.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->
<!-- If a PR has no impact on players (i.e. a code refactor that does not change functionality) then the entire Changelog heading and contents can be removed. -->

CinnamonSnowball commented:
Some usability nitpicks I found while testing:
-​ It's really hard to move the popup around, there's only a couple pixels of the border that you can click-drag to move the window
-​ Clicking on the target channel button to switch it drops focus on the text field, meaning you have to click it before you can actually start typing again, would be neat if the focus could be automatically transferred back to the field
-​ `;` Common comms don't seem to be recognized as a valid channel switch like, say, `.m` is; wonder if that's due to some limitations?
-​ While pressing Backspace to erase selected text works fine, Ctrl+Backspace appears to ignore the selected text and just deletes the word before the caret's current position
-​ Ctrl+Delete to forward-erase a word looks unsupported, sadly, does nothing whatsoever
-​ It's impossible to change the selected channel while there's any message typed in, requiring you to either delete all of the text, press Backspace again, then type the intended channel key and start again, or manually click the target channel button to swap to `Say` and type the intended channel key at the front. Bit of a regression, I feel it'd be good to just always allow erasing channel selection by Backspacing the start of the message

Contrabang commented on code/__HELPERS/text.dm:
```diff
@@ -269,7 +269,9 @@
 
 //Returns a string with reserved characters and spaces before the first word and after the last word removed.
-/proc/trim(text)
+/proc/trim(text, max_length)
+	if(max_length)
+		text = copytext_char(text, 1, max_length)
```
IMO, this should not be included as part of this function, it should stay simple, and if someone needs to set a max length, they can do that.

Contrabang commented on code/modules/tgui/tgui_panel/tgui_panel.dm:
```diff
@@ -13,9 +13,9 @@
 	var/initialized_at
 
-/datum/tgui_panel/New(client/client)
+/datum/tgui_panel/New(client/client, id)
```
do we want a default value for "id" here?

S34NW replied:
No, but we do want to prevent making panels without an ID, i'll add safeties

Contrabang commented on code/modules/tgui_input/say_modal/tgui_say_modal.dm:
```diff
@@ -0,0 +1,137 @@
+	/// The user who opened the window
+	var/client/client
+	/// Injury phrases to blurt out
+	var/list/hurt_phrases = list("GACK!", "GLORF!", "OOF!", "AUGH!", "OW!", "URGH!", "HRNK!")
```
Some of these are kind of silly, and I'm not entirely against them, but I think it would be good to have more realistic ones. ("AGH!", "ARGH!!", "ACK!!", "OUF!" are some I've seen on other servers).

(Also, should this list be static?)

Contrabang commented on code/modules/tgui_input/say_modal/tgui_say_modal.dm:
```diff
@@ -0,0 +1,137 @@
+	/// Injury phrases to blurt out
+	var/list/hurt_phrases = list("GACK!", "GLORF!", "OOF!", "AUGH!", "OW!", "URGH!", "HRNK!")
+	/// Max message length
+	var/max_length = MAX_MESSAGE_LEN
```
Does this ever change? otherwise we can get rid of this as a variable and just use the define

Contrabang commented on code/modules/tgui_input/say_modal/tgui_say_typing.dm:
```diff
@@ -0,0 +1,50 @@
+/mob/set_stat(new_stat)
+	. = ..()
+	if(.)
+		set_typing_indicator(FALSE)
+
+/mob/Logout()
+	set_typing_indicator(FALSE)
+	return ..()
```
IMO it would be better to add these to their parent procs, or at least the logout one, since `mob/Logout` has its own file entirely for that proc.

Contrabang reviewed (changes requested):
Some DM-side review

matttheficus commented:
Can there be examples of the features in-game within the PR itself once theyre up and working?

S34NW commented:
> Can there be examples of the features in-game within the PR itself once theyre up and working?

There are. Click the dropdown at the bottom of the "What Does This PR Do" section. They are from the TG PR but work the same here.

S34NW commented:
> Some usability nitpicks I found while testing:
> 
>     * It's really hard to move the popup around, there's only a couple pixels of the border that you can click-drag to move the window; personally I also feel it's a bit small and not the easiest to see, but that might be just me

![image](https://github.com/ParadiseSS13/Paradise/assets/12197162/ed1ed918-4bc6-45ef-9649-93eb3a0abf36)
The space to the left of the channel button is all draggable, same with a section at the end of the message. Perhaps it would be better if we added a little handle element to the left though?


>     * Clicking on the target channel button to switch it drops focus on the text field, meaning you have to click it before you can actually start typing again, would be neat if the focus could be automatically transferred back to the field

I've implemented this now, but due to limitations in my first iteration, it can only set focus back to the end of the string.

>     * `;` Common comms don't seem to be recognized as a valid channel switch like, say, `.m` is; wonder if that's due to some limitations?

Working on adding that now.

>     * While pressing Backspace to erase selected text works fine, Ctrl+Backspace appears to ignore the selected text and just deletes the word before the caret's current position

This will be tweaked

>     * Ctrl+Delete to forward-erase a word looks unsupported, sadly, does nothing whatsoever

This will be added

>     * It's impossible to change the selected channel while there's any message typed in, requiring you to either delete all of the text, press Backspace again, then type the intended channel key and start again, or manually click the target channel button to swap to `Say` and type the intended channel key at the front. Bit of a regression, I feel it'd be good to just always allow erasing channel selection by Backspacing the start of the message

From testing on my end, if you have a message and go back to enter a radio key (.s, .m, etc.) at the start it does change the radio key. That being said, I have changed the behaviour so that backspace at the start of the message now resets you to the base "say" channel.

Contrabang commented on code/modules/tgui_input/say_modal/tgui_say_modal.dm:
```diff
@@ -0,0 +1,135 @@
+/datum/tgui_say/proc/load()
+	window_open = FALSE
+	winset(client, "tgui_say", "pos=848,500;size=231,30;is-visible=0;")
+	window.send_message("props", list(
+		lightMode = (client.prefs.toggles2 & PREFTOGGLE_2_ENABLE_TGUI_SAY_LIGHT_MODE),
+		maxLength = MAX_MESSAGE_LEN,
+	))
```
I think these are keys in an associative list, right? If so, we should clarify that
```suggestion
	window.send_message("props", list(
		"lightMode" = (client.prefs.toggles2 & PREFTOGGLE_2_ENABLE_TGUI_SAY_LIGHT_MODE),
		"maxLength" = MAX_MESSAGE_LEN,
	))
```

Contrabang commented on code/modules/tgui_input/say_modal/tgui_say_modal.dm:
```diff
@@ -0,0 +1,135 @@
+ */
+/datum/tgui_say/proc/toggle_dark_light_mode()
+	window.send_message("props", list(
+		lightMode = (client.prefs.toggles2 & PREFTOGGLE_2_ENABLE_TGUI_SAY_LIGHT_MODE),
```
ditto
```suggestion
		"lightMode" = (client.prefs.toggles2 & PREFTOGGLE_2_ENABLE_TGUI_SAY_LIGHT_MODE),
```

Contrabang reviewed (commented):
I think these are keys in an associative list, right? If so, we should clarify that

S34NW commented:
![image](https://github.com/ParadiseSS13/Paradise/assets/12197162/ffe0e6f8-8974-4003-9043-d7ca20ffc5bb)
Added drag handle indicators to make it easier to know how you can drag the box.

Example of these from VSC: 
![image](https://github.com/ParadiseSS13/Paradise/assets/12197162/975a417f-24ec-49d3-84b3-c4a1444d6446)

DogeDogIs commented:
s34n 

you're the best

Qwertytoforty commented:
Show/Hide Typing Indicator is broken, you always show a typing indicator

Alecksohs commented:
Whisper doesn't have a menu yet.

S34NW commented:
> Whisper doesn't have a menu yet.

It does :)

S34NW commented:
> Show/Hide Typing Indicator is broken, you always show a typing indicator

How bizzare, we still use the old code for this. Will investigate

Henri215 commented:
The issue https://github.com/ParadiseSS13/Paradise/issues/24074 is related to this PR.

DGamerL commented on code/modules/tgui_input/say_modal/tgui_say_modal.dm:
```diff
@@ -0,0 +1,128 @@
+/** Assigned say modal of the client */
+/client/var/datum/tgui_say/tgui_say
```
Move this to `client_defines` please

lewcc commented on code/datums/keybindings/communication_keybinds.dm:
```diff
@@ -0,0 +1,74 @@
+		to_chat(C, "<span class='danger'>OOC is globally muted.</span>", MESSAGE_TYPE_WARNING)
+		return
+
+	if(!GLOB.dooc_enabled && (C.mob.stat == DEAD))
```
```suggestion
	if(!GLOB.dooc_enabled && C.mob.stat == DEAD)
```

lewcc reviewed (commented):
two lil things

Burzah reviewed (approved):
Tentative approval for the typescript and the styling.

AyIong reviewed (commented):
Honestly didn't notice any snags, probably due to not having much experience, but most likely the rest is just already fixed xdd

S34NW commented on code/__DEFINES/preferences_defines.dm:
```diff
@@ -1,73 +1,74 @@
+#define PREFTOGGLE_CHAT_PRAYER					16
+#define PREFTOGGLE_CHAT_RADIO					32
+// #define PREFTOGGLE_AZERTY					64 // obsolete
+#define PREFTOGGLE_CHAT_DEBUGLOGS 				28
```
```suggestion
#define PREFTOGGLE_CHAT_DEBUGLOGS 				128
```

AyIong reviewed (approved):
LGTM
Will port it to SierraBay 🐀 (Maybe)

farie82 commented on code/modules/tgui_input/say_modal/tgui_say_speech.dm:
```diff
@@ -0,0 +1,55 @@
+	if(!payload?["channel"] || !payload["entry"])
+		CRASH("[usr] entered in a null payload to the chat window.")
+	if(length(payload["entry"]) > MAX_MESSAGE_LEN)
+		CRASH("[usr] has entered more characters than allowed into a TGUI-Say")
```
Is this in any way abusable somehow? Or is it prechecked by other code and this is more a developer thing?

S34NW replied:
This was handled in TS as a first line of defense. But after discussion on discord I've made the process more robust.
